taskcat/amiupdater.py

```diff
@@ -0,0 +1,409 @@
+        # The RegionalCodename instance can access this value. That's important for threading
+        # the API queries - which we do.
+        cnfilter = TemplateClass.deep_get(Config.raw_dict, "global/AMIs/{}".format(self.cn))
+        if self._filters:
+            cnfilter = self._filters
+        if cnfilter:
+            self.filters = [{'Name':k, 'Values': [v]} for k, v in cnfilter.items()]
+            self.filters.append({'Name':'state', 'Values':['available']})
+        if not self.filters:
+            return None
+        return True
+
+    def regions(self):
+        return list(self._regions)
+
+    @classmethod
+    def objects(cls):
+        return list(cls._objs.values())
+
+    @classmethod
+    def fetch_latest_amis(cls):
+        """Fetches AMI IDs from AWS"""
+        # This is a wrapper for the threaded run.
+        # Create a ThreadPool, size is the number of regions.
+
+        if len(RegionalCodename.objects()) == 0:
+            raise AMIUpdaterException("No AMI filters were found. Nothing to fetch from the EC2 API.")
+
+        pool = ThreadPool(len(TemplateClass.regions()))
+        # For reach RegionalCodename that we've generated....
+        pool.map(cls._per_rcn_ami_fetch, RegionalCodename.objects())
+
+    @classmethod
+    def _per_rcn_ami_fetch(cls, rcn):
+        rcn.results = AMIUpdater.client_factory.get('ec2', rcn.region).describe_images(Filters=rcn.filters)['Images']
+
+    @classmethod
+    def parse_api_results(cls):
+        raw_ami_names = {}
+        region_codename_result_list = []
+
+        # For each RegionalCodename.
+        #     Create a Dictionary like so:
+        #     CODENAME
+        #       - REGION_NAME
+        #           [{RAW_API_RESULTS_1}, {RAW_API_RESULTS_2}, {RAW_API_RESULTS_N}]
+        for rcn in RegionalCodename.objects():
+            if rcn.cn in raw_ami_names.keys():
+                raw_ami_names[rcn.cn][rcn.region] = [
+                    APIResultsData(rcn.cn, x['ImageId'],
+                                   int(datetime.datetime.strptime(x['CreationDate'],
+                                                                  "%Y-%m-%dT%H:%M:%S.%fZ").timestamp()), rcn.region)
+                    for x in rcn.results]
+            else:
+                raw_ami_names.update({rcn.cn:
+                                          {rcn.region: [APIResultsData(rcn.cn, x['ImageId'],
+                                                                       int(datetime.datetime.strptime(x['CreationDate'],
+                                                                                                      "%Y-%m-%dT%H:%M:%S.%fZ").timestamp()), rcn.region)
+                                                        for x in rcn.results]}})
+
+        for codename, regions in raw_ami_names.items():
+            for region, results_list in regions.items():
+                latest_ami = sorted(results_list, reverse=True)[0]
```

Comment:
should check the length of the list first to determine if there are any results and avoid this exception:

```
amiupdater ~/quickstart-amazon-eks/templates/amazon-eks-nodegroup.template.yaml
[INFO   ] : Created all filters necessary for the API calls
[INFO   ] : Latest AMI IDs fetched
Traceback (most recent call last):
  File "/Users/jmmccon/taskcat/venv-pc/bin/amiupdater", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/Users/jmmccon/taskcat/bin/amiupdater", line 38, in <module>
    a.update_amis()
  File "/Users/jmmccon/taskcat/taskcat/amiupdater.py", line 398, in update_amis
    Codenames.parse_api_results()
  File "/Users/jmmccon/taskcat/taskcat/amiupdater.py", line 155, in parse_api_results
    latest_ami = sorted(results_list, reverse=True)[0]
IndexError: list index out of range
```